apps/meeting/src/providers/DataMessagesProvider/index.tsx

```diff
@@ -0,0 +1,103 @@
+
+export const DataMessagesProvider: FC = ({ children }) => {
+  const { localUserName } = useAppState();
+  const meetingManager = useMeetingManager();
+  const audioVideo = useAudioVideo();
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  useEffect(() => {
+    if (!audioVideo) {
+      return;
+    }
+    audioVideo.realtimeSubscribeToReceiveDataMessage('ChimeComponentLibraryDataMessage', handler);
+    return () => {
+      audioVideo.realtimeUnsubscribeFromReceiveDataMessage('ChimeComponentLibraryDataMessage');
+    };
+  }, [audioVideo]);
+
+  const handler = useCallback(
+    (dataMessage: DataMessage) => {
+      if (!dataMessage.throttled) {
+        const isSelf =
+          dataMessage.senderAttendeeId === meetingManager.meetingSession?.configuration.credentials?.attendeeId;
+        if (isSelf) {
+          dispatch({
+            type: DataMessagesActionType.ADD,
+            payload: {
+              message: new TextDecoder().decode(dataMessage.data),
+              senderAttendeeId: dataMessage.senderAttendeeId,
+              timestamp: dataMessage.timestampMs,
+              senderName: dataMessage.senderExternalUserId,
+              isSelf: true,
+            },
+          });
+        } else {
+          const data = dataMessage.json();
+          dispatch({
+            type: DataMessagesActionType.ADD,
+            payload: {
+              message: data.message,
+              senderAttendeeId: dataMessage.senderAttendeeId,
+              timestamp: dataMessage.timestampMs,
+              senderName: data.senderName,
+              isSelf: false,
+            },
+          });
+        }
+      } else {
+        console.warn('DataMessage is throttled. Please resend');
+      }
+    },
+    [meetingManager]
+  );
+
+  const sendMessage = useCallback(
+    (message: string) => {
+      if (!meetingManager || !audioVideo) {
+        return;
+      }
+      const DATA_MESSAGE_LIFETIME_MS = 30000;
+      const payload = { message, senderName: localUserName };
+      audioVideo.realtimeSendDataMessage('ChimeComponentLibraryDataMessage', payload, DATA_MESSAGE_LIFETIME_MS);
+      handler(
+        new DataMessage(
+          Date.now(),
+          'ChimeComponentLibraryDataMessage',
+          new TextEncoder().encode(message),
+          meetingManager.meetingSession?.configuration?.credentials?.attendeeId!,
+          localUserName
+        )
+      );
+    },
+    [meetingManager, audioVideo]
+  );
+
+  const value = {
+    sendMessage,
+    messages: state.messages,
+  };
+  return <DataMessagesStateContext.Provider value={value}>{children}</DataMessagesStateContext.Provider>;
+};
+
+export const useDataMessages = (): {
+  sendMessage: (message: string) => void;
+  messages: ChatDataMessage[];
+} => {
+  const context = useContext(DataMessagesStateContext);
+  if (!context) {
+    throw new Error('Use useDataMessages hook inside DataMessagesProvider');
+  }
```

Comment:
Does this need to check whether we're in a meeting provider context too?

Reply:
Good point! Added a check for meeting manager.